File: backend/sandbox.py
import os
import re
import uuid
import shutil
import subprocess
import asyncio
from pathlib import Path

# Registry-qualified ref for the packaged app (set via env); local name for dev.
SANDBOX_IMAGE = os.environ.get("CPPBOX_SANDBOX_IMAGE", "cpp-sandbox:latest")
WORKDIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "workdir")


def _runtime() -> str:
    """Container CLI: prefer podman (rootless), fall back to docker."""
    if shutil.which("podman"):
        return "podman"
    if shutil.which("docker"):
        return "docker"
    return "podman"  # let `run` fail with a clear message if neither exists


_RUNTIME = _runtime()


def ensure_sandbox_image() -> None:
    """Best-effort prefetch of the sandbox image. Registry-qualified refs
    auto-pull on `run` anyway; this just avoids a first-compile stall.
    Errors (offline, etc.) are ignored — the compile path reports clearly."""
    try:
        if subprocess.run([_RUNTIME, "image", "inspect", SANDBOX_IMAGE],
                          capture_output=True).returncode != 0:
            subprocess.run([_RUNTIME, "pull", SANDBOX_IMAGE], capture_output=True)
    except Exception:
        pass


class CompileResult:
    def __init__(self, success: bool, output: str, binary_path: str | None = None):
        self.success = success
        self.output = output
        self.binary_path = binary_path


class RunResult:
    def __init__(self, output: str, exit_code: int, timed_out: bool = False):
        self.output = output
        self.exit_code = exit_code
        self.timed_out = timed_out


async def compile_files(files: list[dict], std: str = "c++17") -> CompileResult:
    """Compile multiple files together. Each file: {name, content}."""
    job_id = uuid.uuid4().hex[:12]
    job_dir = Path(WORKDIR) / job_id
    job_dir.mkdir(parents=True, exist_ok=True)
    os.chmod(job_dir, 0o777)

    # Write all files
    for f in files:
        p = job_dir / f["name"]
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(f["content"])
        p.chmod(0o666)

    # Determine source file list
    all_sources = " ".join(
        f["name"] for f in files if f["name"].endswith((".cpp", ".cc", ".cxx", ".c"))
    )
    if not all_sources:
        all_sources = " ".join(f["name"] for f in files)

    # include every directory (relative) so headers in subdirs resolve
    inc_dirs = {"."}
    for f in files:
        parent = os.path.dirname(f["name"])
        while parent:
            inc_dirs.add(parent)
            parent = os.path.dirname(parent)
    inc_flags = " ".join(f"-I{d}" for d in sorted(inc_dirs))

    binary = job_dir / "a.out"

    try:
        proc = await asyncio.create_subprocess_exec(
            _RUNTIME, "run", "--rm",
            "--network", "none",
            "--memory", "512m",
            "--cpus", "2",
            "--pids-limit", "50",
            "--read-only",
            "--security-opt", "label=disable",
            "-v", f"{job_dir}:/home/sandbox/work:rw",
            "-w", "/home/sandbox/work",
            SANDBOX_IMAGE,
            "sh", "-c",
            f"clang++ -std={std} -O2 -Wall -Wextra -pedantic -fcolor-diagnostics {inc_flags} {all_sources} -o a.out 2>&1",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=60)
        output = (stdout + stderr).decode("utf-8", errors="replace")

        if proc.returncode == 0 and binary.exists():
            return CompileResult(success=True, output=output, binary_path=str(binary))
        else:
            return CompileResult(success=False, output=output or "Compilation failed (no output)")

    except asyncio.TimeoutError:
        return CompileResult(success=False, output="Compilation timed out (60s)")
    except Exception as e:
        return CompileResult(success=False, output=f"Compilation error: {e}")


async def run_binary(binary_path: str, stdin: str = "", timeout: int = 15) -> RunResult:
    """Run a compiled binary in an isolated Docker container."""
    job_dir = Path(binary_path).parent
    binary_name = Path(binary_path).name

    try:
        proc = await asyncio.create_subprocess_exec(
            _RUNTIME, "run", "--rm", "-i",
            "--network", "none",
            "--memory", "256m",
            "--cpus", "1",
            "--pids-limit", "30",
            "--read-only",
            "--security-opt", "label=disable",
            "--cap-drop", "ALL",
            "--security-opt", "no-new-privileges",
            "-v", f"{job_dir}:/home/sandbox/work:ro",
            "-w", "/home/sandbox/work",
            SANDBOX_IMAGE,
            "sh", "-c",
            f"timeout {timeout} ./{binary_name} 2>&1",
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(
            proc.communicate(input=stdin.encode() if stdin else None),
            timeout=timeout + 10,
        )

        output = (stdout + stderr).decode("utf-8", errors="replace")
        timed_out = proc.returncode == 124
        return RunResult(output=output, exit_code=proc.returncode or 0, timed_out=timed_out)

    except asyncio.TimeoutError:
        return RunResult(output="Execution timed out.", exit_code=-1, timed_out=True)
    except Exception as e:
        return RunResult(output=f"Execution error: {e}", exit_code=-1)


async def compile_and_run_files(files: list[dict], stdin: str = "", std: str = "c++17") -> dict:
    """Compile multiple files then run."""
    compile_result = await compile_files(files, std=std)

    if not compile_result.success:
        if compile_result.binary_path:
            _cleanup(Path(compile_result.binary_path).parent)
        return {
            "ok": False,
            "stage": "compile",
            "compile_output": compile_result.output,
            "run_output": "",
        }

    try:
        run_result = await run_binary(compile_result.binary_path, stdin)
    finally:
        _cleanup(Path(compile_result.binary_path).parent)

    return {
        "ok": run_result.exit_code == 0,
        "stage": "run",
        "compile_output": compile_result.output,
        "run_output": run_result.output,
        "exit_code": run_result.exit_code,
        "timed_out": run_result.timed_out,
    }


async def format_code(code: str, style: str = "LLVM") -> str:
    """Format C++ code using clang-format from the host."""
    try:
        proc = await asyncio.create_subprocess_exec(
            "clang-format", f"--style={style}",
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(
            proc.communicate(input=code.encode()),
            timeout=10,
        )
        if proc.returncode == 0:
            return stdout.decode("utf-8", errors="replace")
        return code  # return original on failure
    except Exception:
        return code


_DIAG_RE = re.compile(r"^(?P<file>[^:]+):(?P<line>\d+):(?P<col>\d+):\s*(?P<sev>error|warning|note|fatal error):\s*(?P<msg>.*)$")


async def check_syntax(files: list[dict], std: str = "c++17", entry: str | None = None) -> list[dict]:
    """Parse-only check on the HOST (no Docker): fast, never executes code.
    Returns [{file, line, col, severity, message}]."""
    job_id = uuid.uuid4().hex[:12]
    job_dir = Path(WORKDIR) / ("chk_" + job_id)
    job_dir.mkdir(parents=True, exist_ok=True)

    for f in files:
        p = job_dir / f["name"]
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(f["content"])

    if entry is None:
        cpp = [f["name"] for f in files if f["name"].endswith((".cpp", ".cc", ".cxx"))]
        entry = cpp[0] if cpp else files[0]["name"]

    # include every directory (absolute) so headers in subdirs resolve
    inc_dirs = {str(job_dir)}
    for f in files:
        parent = os.path.dirname(f["name"])
        while parent:
            inc_dirs.add(str(job_dir / parent))
            parent = os.path.dirname(parent)
    inc_args = []
    for d in sorted(inc_dirs):
        inc_args += ["-I", d]

    try:
        proc = await asyncio.create_subprocess_exec(
            "clang++", f"-std={std}", "-fsyntax-only", "-fno-color-diagnostics",
            "-Wall", "-Wextra", *inc_args,
            str(job_dir / entry),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=20)
        text = (stdout + stderr).decode("utf-8", errors="replace")
    except Exception:
        _cleanup(job_dir)
        return []

    diags = []
    for line in text.splitlines():
        m = _DIAG_RE.match(line)
        if not m:
            continue
        diags.append({
            "file": os.path.basename(m.group("file")),
            "line": int(m.group("line")),
            "col": int(m.group("col")),
            "severity": "error" if "error" in m.group("sev") else "warning" if "warning" in m.group("sev") else "info",
            "message": m.group("msg").strip(),
        })

    _cleanup(job_dir)
    return diags


def _cleanup(job_dir: Path):
    """Remove temp directory."""
    try:
        import shutil
        if job_dir.exists():
            shutil.rmtree(job_dir, ignore_errors=True)
    except Exception:
        pass

File: crates/cppbox-core/src/sandbox.rs
//! Isolated compile/run via podman (preferred) or docker, plus host-side
//! clang-format and clang++ syntax check. Mirrors backend/sandbox.py.
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::sync::OnceLock;
use std::time::Duration;

use serde_json::{json, Value};
use tokio::process::Command;

pub fn sandbox_image() -> String {
    std::env::var("CPPBOX_SANDBOX_IMAGE").unwrap_or_else(|_| "cpp-sandbox:latest".into())
}

/// Compile with debug symbols (-g -O0) for gdb. Returns (binary, job_dir).
pub async fn compile_debug(root: &Path, files: &[File], std: &str) -> Result<(PathBuf, PathBuf), String> {
    let dir = job_dir(root);
    write_sources(&dir, files);
    let sources = source_list(files);
    let inc = include_flags(files);
    let cmd_str = format!(
        "clang++ -g -O0 -fno-omit-frame-pointer -std={std} -Wall -Wextra -fcolor-diagnostics {inc} {src} -o a.out 2>&1",
        src = sources.join(" ")
    );
    let abs = dir.canonicalize().map_err(|e| e.to_string())?;
    let mut c = Command::new(runtime());
    c.args(["run", "--rm", "--network", "none", "--memory", "512m", "--cpus", "2",
            "--pids-limit", "50", "--read-only", "--security-opt", "label=disable"])
        .arg("-v").arg(format!("{}:/home/sandbox/work:rw", abs.display()))
        .args(["-w", "/home/sandbox/work"])
        .arg(sandbox_image())
        .args(["sh", "-c", &cmd_str])
        .stdout(Stdio::piped()).stderr(Stdio::piped());
    let out = match tokio::time::timeout(Duration::from_secs(60), c.output()).await {
        Ok(Ok(o)) => o,
        Ok(Err(e)) => return Err(format!("compile error: {e}")),
        Err(_) => return Err("Compilation timed out (60s)".into()),
    };
    let mut text = String::from_utf8_lossy(&out.stdout).into_owned();
    text.push_str(&String::from_utf8_lossy(&out.stderr));
    let binary = dir.join("a.out");
    if out.status.success() && binary.exists() {
        Ok((binary, dir))
    } else {
        Err(if text.trim().is_empty() { "Compilation failed".into() } else { text })
    }
}

/// Pick the container CLI once: podman (rootless) preferred, docker fallback.
pub fn runtime() -> &'static str {
    static RT: OnceLock<&'static str> = OnceLock::new();
    *RT.get_or_init(|| {
        for c in ["podman", "docker"] {
            if std::process::Command::new(c)
                .arg("--version")
                .stdout(Stdio::null())
                .stderr(Stdio::null())
                .status()
                .is_ok()
            {
                return c;
            }
        }
        "podman"
    })
}

pub struct File {
    pub name: String,
    pub content: String,
}

struct CompileResult {
    success: bool,
    output: String,
    binary: Option<PathBuf>,
}

struct RunResult {
    output: String,
    exit_code: i32,
    timed_out: bool,
}

fn job_dir(root: &Path) -> PathBuf {
    let id = uuid::Uuid::new_v4().simple().to_string();
    let d = root.join("workdir").join(&id[..12]);
    std::fs::create_dir_all(&d).ok();
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&d, std::fs::Permissions::from_mode(0o777)).ok();
    }
    d
}

fn write_sources(dir: &Path, files: &[File]) {
    for f in files {
        let p = dir.join(&f.name);
        if let Some(parent) = p.parent() {
            std::fs::create_dir_all(parent).ok();
        }
        std::fs::write(&p, &f.content).ok();
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            std::fs::set_permissions(&p, std::fs::Permissions::from_mode(0o666)).ok();
        }
    }
}

fn source_list(files: &[File]) -> Vec<String> {
    let exts = ["cpp", "cc", "cxx", "c"];
    let mut s: Vec<_> = files
        .iter()
        .filter(|f| exts.iter().any(|e| f.name.ends_with(e)))
        .map(|f| f.name.clone())
        .collect();
    if s.is_empty() {
        s = files.iter().map(|f| f.name.clone()).collect();
    }
    s
}

fn include_flags(files: &[File]) -> String {
    let mut dirs = vec![".".to_string()];
    for f in files {
        let parent = Path::new(&f.name).parent();
        let mut p = parent.map(|x| x.to_path_buf());
        while let Some(pp) = p {
            let s = pp.to_string_lossy().into_owned();
            if !s.is_empty() && !dirs.contains(&s) {
                dirs.push(s);
            }
            p = pp.parent().map(|x| x.to_path_buf());
        }
    }
    dirs.iter().map(|d| format!("-I{}", d)).collect::<Vec<_>>().join(" ")
}

async fn compile_files(root: &Path, files: &[File], std: &str) -> CompileResult {
    let dir = job_dir(root);
    write_sources(&dir, files);
    let sources = source_list(files);
    let inc = include_flags(files);
    let cmd_str = format!(
        "clang++ -std={} -O2 -Wall -Wextra -pedantic -fcolor-diagnostics {} {} -o a.out 2>&1",
        std, inc, sources.join(" ")
    );
    let abs = match dir.canonicalize() {
        Ok(a) => a,
        Err(e) => return CompileResult { success: false, output: format!("job dir error: {e}"), binary: None },
    };

    let mut c = Command::new(runtime());
    c.args(["run", "--rm", "--network", "none", "--memory", "512m", "--cpus", "2",
            "--pids-limit", "50", "--read-only", "--security-opt", "label=disable"])
        .arg("-v").arg(format!("{}:/home/sandbox/work:rw", abs.display()))
        .args(["-w", "/home/sandbox/work"])
        .arg(sandbox_image())
        .args(["sh", "-c", &cmd_str])
        .stdout(Stdio::piped()).stderr(Stdio::piped());

    let output = match tokio::time::timeout(Duration::from_secs(60), c.output()).await {
        Ok(Ok(o)) => o,
        Ok(Err(e)) => return CompileResult { success: false, output: format!("Compilation error: {e}"), binary: None },
        Err(_) => return CompileResult { success: false, output: "Compilation timed out (60s)".into(), binary: None },
    };
    let mut text = String::from_utf8_lossy(&output.stdout).into_owned();
    text.push_str(&String::from_utf8_lossy(&output.stderr));
    let binary = dir.join("a.out");
    let success = output.status.success() && binary.exists();
    let binary_path = if success { Some(binary) } else { None };
    let output_text = if text.trim().is_empty() { "Compilation failed (no output)".into() } else { text };
    CompileResult { success, output: output_text, binary: binary_path }
}

async fn run_binary(binary: &Path, stdin: &str, timeout: u64) -> RunResult {
    let dir = match binary.parent() {
        Some(d) => d.to_path_buf(),
        None => return RunResult { output: "invalid binary path".into(), exit_code: -1, timed_out: false },
    };
    let name = binary.file_name().and_then(|n| n.to_str()).unwrap_or("a.out");
    let abs = match dir.canonicalize() {
        Ok(a) => a,
        Err(e) => return RunResult { output: format!("Execution error: {e}"), exit_code: -1, timed_out: false },
    };
    let cmd_str = format!("timeout {timeout} ./{name} 2>&1");

    let mut c = Command::new(runtime());
    c.args(["run", "--rm", "-i", "--network", "none", "--memory", "256m", "--cpus", "1",
            "--pids-limit", "30", "--read-only", "--security-opt", "label=disable",
            "--cap-drop", "ALL", "--security-opt", "no-new-privileges"])
        .arg("-v").arg(format!("{}:/home/sandbox/work:ro", abs.display()))
        .args(["-w", "/home/sandbox/work"])
        .arg(sandbox_image())
        .args(["sh", "-c", &cmd_str])
        .stdin(Stdio::piped()).stdout(Stdio::piped()).stderr(Stdio::piped());

    let mut child = match c.spawn() {
        Ok(ch) => ch,
        Err(e) => return RunResult { output: format!("Execution error: {e}"), exit_code: -1, timed_out: false },
    };
    if let Some(mut sin) = child.stdin.take() {
        use tokio::io::AsyncWriteExt;
        let _ = sin.write_all(stdin.as_bytes()).await;
    }

    match tokio::time::timeout(Duration::from_secs(timeout + 10), child.wait_with_output()).await {
        Ok(Ok(o)) => {
            let mut text = String::from_utf8_lossy(&o.stdout).into_owned();
            text.push_str(&String::from_utf8_lossy(&o.stderr));
            let code = o.status.code().unwrap_or(-1);
            let timed_out = code == 124;
            RunResult { output: text, exit_code: code, timed_out }
        }
        Ok(Err(e)) => RunResult { output: format!("Execution error: {e}"), exit_code: -1, timed_out: false },
        Err(_) => RunResult { output: "Execution timed out.".into(), exit_code: -1, timed_out: true },
    }
}

fn cleanup(dir: &Path) {
    let _ = std::fs::remove_dir_all(dir);
}

/// Compile then run. Returns the JSON shape the frontend expects.
pub async fn compile_and_run(root: &Path, files: &[File], stdin: &str, std: &str) -> Value {
    let cr = compile_files(root, files, std).await;
    if !cr.success {
        if let Some(b) = &cr.binary { if let Some(p) = b.parent() { cleanup(p); } }
        return json!({ "ok": false, "stage": "compile", "compile_output": cr.output, "run_output": "" });
    }
    let binary = cr.binary.clone().unwrap();
    let rr = run_binary(&binary, stdin, 15).await;
    if let Some(p) = binary.parent() { cleanup(p); }
    json!({
        "ok": rr.exit_code == 0,
        "stage": "run",
        "compile_output": cr.output,
        "run_output": rr.output,
        "exit_code": rr.exit_code,
        "timed_out": rr.timed_out,
    })
}

/// Host clang-format (no container) — returns original on failure.
pub async fn format_code(code: &str, style: &str) -> String {
    let mut c = Command::new("clang-format");
    c.arg(format!("--style={style}"))
        .stdin(Stdio::piped()).stdout(Stdio::piped()).stderr(Stdio::piped());
    let mut child = match c.spawn() {
        Ok(ch) => ch,
        Err(_) => return code.into(),
    };
    if let Some(mut sin) = child.stdin.take() {
        use tokio::io::AsyncWriteExt;
        let _ = sin.write_all(code.as_bytes()).await;
    }
    match tokio::time::timeout(Duration::from_secs(10), child.wait_with_output()).await {
        Ok(Ok(o)) if o.status.success() => String::from_utf8_lossy(&o.stdout).into_owned(),
        _ => code.into(),
    }
}

/// Host clang++ -fsyntax-only parse; returns diagnostics.
pub async fn check_syntax(root: &Path, files: &[File], std: &str, entry: Option<&str>) -> Vec<Value> {
    let dir = job_dir(root);
    write_sources(&dir, files);

    let entry = entry.map(str::to_string).unwrap_or_else(|| {
        let cpp: Vec<_> = files.iter().filter(|f| {
            [".cpp", ".cc", ".cxx"].iter().any(|e| f.name.ends_with(e))
        }).collect();
        cpp.first().map(|f| f.name.clone()).or_else(|| files.first().map(|f| f.name.clone())).unwrap_or_default()
    });
    if entry.is_empty() {
        cleanup(&dir);
        return vec![];
    }

    // include dirs (absolute) so headers in subdirs resolve
    let mut inc: Vec<String> = vec![dir.display().to_string()];
    for f in files {
        let mut p = Path::new(&f.name).parent().map(|x| dir.join(x));
        while let Some(pp) = p {
            let s = pp.display().to_string();
            if !inc.contains(&s) { inc.push(s); }
            p = pp.parent().map(|x| dir.join(x));
        }
    }
    let mut args: Vec<String> = vec![format!("-std={std}"), "-fsyntax-only".into(), "-fno-color-diagnostics".into(), "-Wall".into(), "-Wextra".into()];
    for d in &inc { args.push("-I".into()); args.push(d.clone()); }
    args.push(dir.join(&entry).display().to_string());

    let mut c = Command::new("clang++");
    c.args(&args).stdout(Stdio::piped()).stderr(Stdio::piped());
    let text = match tokio::time::timeout(Duration::from_secs(20), c.output()).await {
        Ok(Ok(o)) => {
            let mut t = String::from_utf8_lossy(&o.stdout).into_owned();
            t.push_str(&String::from_utf8_lossy(&o.stderr));
            t
        }
        _ => { cleanup(&dir); return vec![]; }
    };

    let re = regex::Regex::new(
        r"^(?P<file>[^:]+):(?P<line>\d+):(?P<col>\d+):\s*(?P<sev>error|warning|note|fatal error):\s*(?P<msg>.*)$",
    ).unwrap();
    let diags: Vec<Value> = text.lines().filter_map(|line| {
        let c = re.captures(line)?;
        let sev = c.name("sev")?.as_str();
        let severity = if sev.contains("error") { "error" }
            else if sev.contains("warning") { "warning" } else { "info" };
        Some(json!({
            "file": Path::new(c.name("file")?.as_str()).file_name().and_then(|n| n.to_str()).unwrap_or(""),
            "line": c.name("line")?.as_str().parse::<i64>().unwrap_or(0),
            "col": c.name("col")?.as_str().parse::<i64>().unwrap_or(0),
            "severity": severity,
            "message": c.name("msg")?.as_str().trim(),
        }))
    }).collect();
    cleanup(&dir);
    diags
}
